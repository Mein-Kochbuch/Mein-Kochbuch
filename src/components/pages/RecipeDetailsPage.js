import React, {useContext} from "react";
import {ScrollView} from "react-native";
import {Redirect, useParams} from 'react-router-native'
import RecipeDetailsHeader from "../recipes/details/RecipeDetailsHeader";
import RecipeDetailsImageGallery from "../recipes/details/RecipeDetailsImageGallery";
import RecipeDetailsActionBar from "../recipes/details/RecipeDetailsActionBar";
import styled from "styled-components/native";
import RecipeDetailsItemComponent from "../recipes/details/RecipeDetailsItemComponent";
import RecipeDetailsPortions from "../recipes/details/RecipeDetailsPortions";
import {AuthContext} from "../../context/AuthProvider";

export default function RecipeDetailsPage({recipeDetails, getRecipeDetailsById, favorizeRecipeById}) {
    const auth = useContext(AuthContext)
    const {id} = useParams()
    const recipe = recipeDetails[id]
    const ingredients = recipe?.zutaten_set.reduce((previousValue, currentValue) => {
        return {zutat: previousValue.zutat.concat("\n").concat(currentValue?.zutat)}
    }).zutat

    getRecipeDetailsById(id)

    const handleFavorize = () => {
        console.log("fav")
        auth.user ?
            favorizeRecipeById(id)
            : <Redirect to={"/login"}/>

    }

    return (
        <ScrollView>
            <RecipeDetailsHeader title={recipe?.title} owner={recipe?.owner}/>
            <RecipeDetailsImageGallery images={recipe?.image_set}/>
            <RecipeDetailsActionBar
                avgRating={recipe?.avg_rating}
                ratingCount={recipe?.rating_count}
                ownRating={recipe?.rating}
                favorite={recipe?.favorite}
                handleFavorize={handleFavorize}/>
            <RowWrapper>
                <ItemStyled content={"Duration: " + recipe?.dauer + " min"}/>
                <ItemStyled content={"Difficulty: " + recipe?.difficulty.difficulty}/>
                <PortionsStyled content={"Portions: " + recipe?.portionen}/>
            </RowWrapper>
            <ItemWrapper content={"Ingredients:\n" + ingredients}/>
            <ItemWrapper content={"Instructions: \n" + recipe?.anleitung}/>
        </ScrollView>
    )
}

const PortionsStyled = styled(RecipeDetailsPortions)`
  width: 49%;
  margin-top: 12px;
  border-radius: 10px;
  border-width: 1px;
  border-color: #fff;
`

const ItemStyled = styled(RecipeDetailsItemComponent)`
  width: 49%;
  justify-content: center;
  border-radius: 10px;
  border-width: 1px;
  border-color: #fff;
`

const ItemWrapper = styled(RecipeDetailsItemComponent)`
  margin: 12px;
  justify-content: center;
  border-radius: 10px;
  border-width: 1px;
  border-color: #fff;
`

const RowWrapper = styled.View`
  margin: 12px;
  gap: 12px;
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  flex-wrap: wrap;
`
