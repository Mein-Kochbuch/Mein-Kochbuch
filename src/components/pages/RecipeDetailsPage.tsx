import React, {useContext} from "react";
import {ScrollView} from "react-native";
import {useHistory, useParams} from 'react-router-native'
import RecipeDetailsHeader from "../recipes/details/RecipeDetailsHeader";
import RecipeDetailsImageGallery from "../recipes/details/RecipeDetailsImageGallery";
import RecipeDetailsActionBar from "../recipes/details/RecipeDetailsActionBar";
import styled from "styled-components/native";
import RecipeDetailsItemComponent from "../recipes/details/RecipeDetailsItemComponent";
import RecipeDetailsPortions from "../recipes/details/RecipeDetailsPortions";
import {AuthContext} from "../../context/AuthProvider";
import {Recipe} from "../../models/Recipe";


interface RecipeDetailsPageProps {
    recipeDetails: Recipe[]
    getRecipeDetailsById: (id: number) => void,
    favorizeRecipeById: (id: number) => void,
    rateRecipeById: (id: number, rating: number) => void
}

export default function RecipeDetailsPage({
                                              recipeDetails,
                                              getRecipeDetailsById,
                                              favorizeRecipeById,
                                              rateRecipeById
                                          }: RecipeDetailsPageProps) {
    const auth = useContext(AuthContext)
    const {id}: { id: string } = useParams()
    const history = useHistory()
    const recipe = recipeDetails[parseInt(id)]
    const ingredients = recipe?.zutaten_set.reduce((previousValue, currentValue) => {
        return {zutat: previousValue.zutat.concat("\n").concat(currentValue?.zutat)}
    }).zutat

    getRecipeDetailsById(parseInt(id))

    const handleFavorize = () => {
        auth.user ?
            favorizeRecipeById(parseInt(id))
            : history.push("/login")
    }

    const handleRating = (rating: number) => {
        auth.user ?
            rateRecipeById(parseInt(id), rating)
            : history.push("/login")
    }

    return (
        <ScrollView>
            <RecipeDetailsHeader title={recipe?.title} owner={recipe?.owner}/>
            <RecipeDetailsImageGallery images={recipe?.image_set}/>
            <RecipeDetailsActionBar
                avgRating={recipe?.avg_rating}
                ratingCount={recipe?.rating_count}
                ownRating={recipe?.rating}
                favorite={recipe?.favorite}
                handleRating={handleRating}
                handleFavorize={handleFavorize}/>
            <RowWrapper>
                <ItemStyled content={"Duration: " + recipe?.dauer + " min"}/>
                <ItemStyled content={"Difficulty: " + recipe?.difficulty.difficulty}/>
                <PortionsStyled content={"Portions: " + recipe?.portionen}/>
            </RowWrapper>
            <ItemWrapper content={"Ingredients:\n" + ingredients}/>
            <ItemWrapper content={"Instructions: \n" + recipe?.anleitung}/>
        </ScrollView>
    )
}

const PortionsStyled = styled(RecipeDetailsPortions)`
  width: 49%;
  margin-top: 12px;
  border-radius: 10px;
  border-width: 1px;
  border-color: #fff;
`

const ItemStyled = styled(RecipeDetailsItemComponent)`
  width: 49%;
  justify-content: center;
  border-radius: 10px;
  border-width: 1px;
  border-color: #fff;
`

const ItemWrapper = styled(RecipeDetailsItemComponent)`
  margin: 12px;
  justify-content: center;
  border-radius: 10px;
  border-width: 1px;
  border-color: #fff;
`

const RowWrapper = styled.View`
  margin: 12px;
  gap: 12px;
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  flex-wrap: wrap;
`
