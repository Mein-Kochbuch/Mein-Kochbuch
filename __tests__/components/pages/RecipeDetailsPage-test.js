import 'react-native';
import React from 'react';

// Note: test renderer must be required after react-native.
import RecipeDetailsPage from '../../../src/components/pages/RecipeDetailsPage';
import {render} from '@testing-library/react-native';
import {Router} from 'react-router-native';

jest.mock('../../../src/components/recipes/details/RecipeDetailsHeader', () => {
  return props => {
    return <div {...props} />;
  };
});

jest.mock(
  '../../../src/components/recipes/details/RecipeDetailsImageGallery',
  () => {
    return props => {
      return <div {...props} />;
    };
  },
);

jest.mock(
  '../../../src/components/recipes/details/RecipeDetailsActionBar',
  () => {
    return props => {
      return <div {...props} />;
    };
  },
);

jest.mock(
  '../../../src/components/recipes/details/RecipeDetailsItemComponent',
  () => {
    return props => {
      return <div {...props} />;
    };
  },
);

jest.mock('react-router-native', () => ({
  ...jest.requireActual('react-router-native'), // use actual for all non-hook parts
  useParams: () => ({
    id: '3',
  }),
}));

describe('RecipeDetailsPage Test', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('RecipeDetailsPage Snapshot Test', () => {
    const history = {
      push: jest.fn(),
      listen: jest.fn(),
      location: {pathname: '/recipes/3'},
    };

    const recipeDetails = {
      3: {
        title: 'recipe-title',
        owner: {name: 'owner-name'},
        image_set: [{image: 'url'}],
        avg_rating: 3.7,
        rating_count: 17,
        rating: 2.5,
        favorite: true,
        dauer: 25,
        difficulty: {difficulty: 'Mittel'},
        portionen: 4,
        zutaten_set: [{zutat: 'mehl'}, {zutat: 'zucker'}],
        anleitung: 'test Anleitung',
      },
    };
    const getRecipeDetailsById = jest.fn();

    const component = render(
      <Router history={history}>
        <RecipeDetailsPage
          recipeDetails={recipeDetails}
          getRecipeDetailsById={getRecipeDetailsById}
        />
      </Router>,
    );

    expect(component.toJSON()).toMatchSnapshot();
    expect(getRecipeDetailsById).toBeCalledWith(3);
  });
});
