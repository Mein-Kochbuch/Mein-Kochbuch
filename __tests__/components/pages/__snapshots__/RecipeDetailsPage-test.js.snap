// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecipeDetailsPage Test 1`] = `
<RCTScrollView>
  <View>
    <div
      owner={
        Object {
          "name": "owner-name",
        }
      }
      title="recipe-title"
    />
    <div
      images={
        Array [
          Object {
            "image": "url",
          },
        ]
      }
    />
    <div
      avgRating={3.7}
      favorite={true}
      handleFavorize={[Function]}
      handleRating={[Function]}
      ownRating={2.5}
      ratingCount={17}
    />
    <View
      style={
        Array [
          Object {
            "display": "flex",
            "flexDirection": "row",
            "flexWrap": "wrap",
            "gap": 12,
            "justifyContent": "space-between",
            "marginBottom": 12,
            "marginLeft": 12,
            "marginRight": 12,
            "marginTop": 12,
          },
        ]
      }
    >
      <div
        content="Duration: 25 min"
        style={
          Array [
            Object {
              "borderColor": "#fff",
              "borderRadius": 10,
              "borderWidth": 1,
              "justifyContent": "center",
              "width": "49%",
            },
          ]
        }
      />
      <div
        content="Difficulty: Mittel"
        style={
          Array [
            Object {
              "borderColor": "#fff",
              "borderRadius": 10,
              "borderWidth": 1,
              "justifyContent": "center",
              "width": "49%",
            },
          ]
        }
      />
      <View
        style={
          Array [
            Object {
              "width": 150,
            },
            Object {
              "borderColor": "#fff",
              "borderRadius": 10,
              "borderWidth": 1,
              "marginTop": 12,
              "width": "49%",
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontSize": 18,
                "paddingBottom": 12,
                "paddingLeft": 12,
                "paddingRight": 12,
                "paddingTop": 12,
              },
            ]
          }
        >
          Portions: 4
        </Text>
      </View>
    </View>
    <div
      content="Ingredients:
mehl
zucker"
      style={
        Array [
          Object {
            "borderColor": "#fff",
            "borderRadius": 10,
            "borderWidth": 1,
            "justifyContent": "center",
            "marginBottom": 12,
            "marginLeft": 12,
            "marginRight": 12,
            "marginTop": 12,
          },
        ]
      }
    />
    <div
      content="Instructions: 
test Anleitung"
      style={
        Array [
          Object {
            "borderColor": "#fff",
            "borderRadius": 10,
            "borderWidth": 1,
            "justifyContent": "center",
            "marginBottom": 12,
            "marginLeft": 12,
            "marginRight": 12,
            "marginTop": 12,
          },
        ]
      }
    />
  </View>
</RCTScrollView>
`;

exports[`RecipeDetailsPage Test RecipeDetailsPage Snapshot Test 1`] = `
<RCTScrollView>
  <View>
    <div
      owner={
        Object {
          "name": "owner-name",
        }
      }
      title="recipe-title"
    />
    <div
      images={
        Array [
          Object {
            "image": "url",
          },
        ]
      }
    />
    <div
      avgRating={3.7}
      favorite={true}
      handleFavorize={[Function]}
      handleRating={[Function]}
      ownRating={2.5}
      ratingCount={17}
    />
    <View
      style={
        Array [
          Object {
            "display": "flex",
            "flexDirection": "row",
            "flexWrap": "wrap",
            "gap": 12,
            "justifyContent": "space-between",
            "marginBottom": 12,
            "marginLeft": 12,
            "marginRight": 12,
            "marginTop": 12,
          },
        ]
      }
    >
      <div
        content="Duration: 25 min"
        style={
          Array [
            Object {
              "borderColor": "#fff",
              "borderRadius": 10,
              "borderWidth": 1,
              "justifyContent": "center",
              "width": "49%",
            },
          ]
        }
      />
      <div
        content="Difficulty: Mittel"
        style={
          Array [
            Object {
              "borderColor": "#fff",
              "borderRadius": 10,
              "borderWidth": 1,
              "justifyContent": "center",
              "width": "49%",
            },
          ]
        }
      />
      <View
        style={
          Array [
            Object {
              "width": 150,
            },
            Object {
              "borderColor": "#fff",
              "borderRadius": 10,
              "borderWidth": 1,
              "marginTop": 12,
              "width": "49%",
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontSize": 18,
                "paddingBottom": 12,
                "paddingLeft": 12,
                "paddingRight": 12,
                "paddingTop": 12,
              },
            ]
          }
        >
          Portions: 4
        </Text>
      </View>
    </View>
    <div
      content="Ingredients:
mehl
zucker"
      style={
        Array [
          Object {
            "borderColor": "#fff",
            "borderRadius": 10,
            "borderWidth": 1,
            "justifyContent": "center",
            "marginBottom": 12,
            "marginLeft": 12,
            "marginRight": 12,
            "marginTop": 12,
          },
        ]
      }
    />
    <div
      content="Instructions: 
test Anleitung"
      style={
        Array [
          Object {
            "borderColor": "#fff",
            "borderRadius": 10,
            "borderWidth": 1,
            "justifyContent": "center",
            "marginBottom": 12,
            "marginLeft": 12,
            "marginRight": 12,
            "marginTop": 12,
          },
        ]
      }
    />
  </View>
</RCTScrollView>
`;
